internal/events/batch_pin_complete.go

```diff
@@ -125,10 +125,15 @@ func (em *eventManager) persistContexts(ctx context.Context, batchPin *blockchai
 	log.L(ctx).Warnf("Batch insert of pins failed - assuming replay and performing upserts: %s", err)
 
 	// Fall back to an upsert
+	// Use a different context to not re-use DB TX from a group run
+	// We only get here if it fails so this would never work in the same DB TX
+	upsertCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
 	for _, pin := range pins {
-		if err := em.database.UpsertPin(ctx, pin); err != nil {
+		if err := em.database.UpsertPin(upsertCtx, pin); err != nil {
```

Comment:
Seems like a very bad idea to intentionally do this outside the database transaction.

Reply:
The batch insert above should be non-fatal (should be configured to return an empty result on conflict)